File: artifacts/api-server/seed.py
"""
Seed the database with demo data for RxPOS.
Run: python3 artifacts/api-server/seed.py
"""
import asyncio
import os
import sys
import uuid

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker, AsyncSession
from urllib.parse import urlparse, urlencode, parse_qs, urlunparse


def build_async_url(db_url: str) -> str:
    if db_url.startswith("postgresql://"):
        db_url = db_url.replace("postgresql://", "postgresql+asyncpg://", 1)
    elif db_url.startswith("postgres://"):
        db_url = db_url.replace("postgres://", "postgresql+asyncpg://", 1)
    parsed = urlparse(db_url)
    params = parse_qs(parsed.query, keep_blank_values=True)
    params.pop("sslmode", None)
    new_query = urlencode({k: v[0] for k, v in params.items()})
    return urlunparse(parsed._replace(query=new_query))


async def seed():
    from app.database import Base
    from app.models import user, branch, medicine, inventory, patient, prescription, sale, supplier, insurance, notification, settings

    db_url = os.environ.get("DATABASE_URL", "")
    if not db_url:
        print("ERROR: DATABASE_URL not set")
        return

    engine = create_async_engine(build_async_url(db_url), echo=False)
    SessionLocal = async_sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

    print("Creating tables...")
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
    print("Tables created.")

    async with SessionLocal() as db:
        from sqlalchemy import select
        from app.models.user import User
        from app.models.branch import Branch
        from app.models.medicine import Medicine
        from app.models.patient import Patient
        import bcrypt as _bcrypt

        def hash_pw(pw: str) -> str:
            return _bcrypt.hashpw(pw.encode("utf-8"), _bcrypt.gensalt()).decode("utf-8")

        # Check if already seeded
        existing = await db.execute(select(User).where(User.email == "admin@rxpos.pk"))
        if existing.scalar_one_or_none():
            print("Database already seeded. Skipping.")
            await engine.dispose()
            return

        # Create main branch
        main_branch = Branch(
            name="RxPOS Main Pharmacy",
            code="HQ",
            address="123 Main Street, Karachi",
            phone="+92-21-1234567",
            license_number="DRAP-PH-2024-001",
            is_active=True,
        )
        db.add(main_branch)
        await db.flush()

        branch2 = Branch(
            name="RxPOS North Branch",
            code="NB",
            address="456 North Road, Lahore",
            phone="+92-42-7654321",
            license_number="DRAP-PH-2024-002",
            is_active=True,
        )
        db.add(branch2)
        await db.flush()

        # Create users
        admin = User(
            name="System Admin",
            email="admin@rxpos.pk",
            hashed_password=hash_pw("admin123"),
            role="super_admin",
            is_active=True,
            branch_id=main_branch.id,
        )
        db.add(admin)
        await db.flush()

        cashier = User(
            name="Ali Cashier",
            email="cashier@rxpos.pk",
            hashed_password=hash_pw("cashier123"),
            role="cashier",
            is_active=True,
            branch_id=main_branch.id,
        )
        db.add(cashier)

        pharmacist = User(
            name="Dr. Sara Pharmacist",
            email="pharmacist@rxpos.pk",
            hashed_password=hash_pw("cashier123"),
            role="pharmacist",
            is_active=True,
            branch_id=main_branch.id,
        )
        db.add(pharmacist)

        manager = User(
            name="Ahmed Manager",
            email="manager@rxpos.pk",
            hashed_password=hash_pw("manager123"),
            role="manager",
            is_active=True,
            branch_id=main_branch.id,
        )
        db.add(manager)
        await db.flush()

        # Set branch manager
        main_branch.manager_id = admin.id
        await db.flush()

        # Create medicines
        medicines_data = [
            {"name": "Panadol 500mg", "generic_name": "Paracetamol", "brand": "GSK", "category": "Analgesics", "form": "Tablet", "strength": "500mg", "unit": "Strip", "pack_size": 10, "manufacturer": "GSK Pakistan", "requires_prescription": False, "mrp": 50.0, "reorder_point": 50, "min_stock_level": 20, "barcode": "6921804900024"},
            {"name": "Augmentin 625mg", "generic_name": "Amoxicillin+Clavulanic Acid", "brand": "GSK", "category": "Antibiotics", "form": "Tablet", "strength": "625mg", "unit": "Strip", "pack_size": 14, "manufacturer": "GSK Pakistan", "requires_prescription": True, "mrp": 450.0, "reorder_point": 30, "min_stock_level": 10},
            {"name": "Ciproxin 500mg", "generic_name": "Ciprofloxacin", "brand": "Bayer", "category": "Antibiotics", "form": "Tablet", "strength": "500mg", "unit": "Strip", "pack_size": 10, "manufacturer": "Bayer AG", "requires_prescription": True, "mrp": 320.0, "reorder_point": 20, "min_stock_level": 10},
            {"name": "Brufen 400mg", "generic_name": "Ibuprofen", "brand": "Abbott", "category": "NSAIDs", "form": "Tablet", "strength": "400mg", "unit": "Strip", "pack_size": 10, "manufacturer": "Abbott Pakistan", "requires_prescription": False, "mrp": 80.0, "reorder_point": 40, "min_stock_level": 15},
            {"name": "Flagyl 400mg", "generic_name": "Metronidazole", "brand": "Sanofi", "category": "Antibiotics", "form": "Tablet", "strength": "400mg", "unit": "Strip", "pack_size": 14, "manufacturer": "Sanofi Pakistan", "requires_prescription": True, "mrp": 120.0, "reorder_point": 25, "min_stock_level": 10},
            {"name": "Glucophage 500mg", "generic_name": "Metformin", "brand": "Merck", "category": "Antidiabetics", "form": "Tablet", "strength": "500mg", "unit": "Strip", "pack_size": 28, "manufacturer": "Merck Pakistan", "requires_prescription": True, "mrp": 280.0, "reorder_point": 30, "min_stock_level": 15},
            {"name": "Norvasc 5mg", "generic_name": "Amlodipine", "brand": "Pfizer", "category": "Antihypertensives", "form": "Tablet", "strength": "5mg", "unit": "Strip", "pack_size": 28, "manufacturer": "Pfizer Pakistan", "requires_prescription": True, "mrp": 350.0, "reorder_point": 20, "min_stock_level": 10},
            {"name": "ORS Sachet", "generic_name": "Oral Rehydration Salts", "brand": "WHO", "category": "OTC", "form": "Sachet", "strength": "N/A", "unit": "Box", "pack_size": 10, "manufacturer": "Various", "requires_prescription": False, "mrp": 40.0, "reorder_point": 50, "min_stock_level": 20},
            {"name": "Vitamin C 500mg", "generic_name": "Ascorbic Acid", "brand": "Sanofi", "category": "Vitamins", "form": "Tablet", "strength": "500mg", "unit": "Bottle", "pack_size": 100, "manufacturer": "Sanofi Pakistan", "requires_prescription": False, "mrp": 220.0, "reorder_point": 30, "min_stock_level": 10},
            {"name": "Omeprazole 20mg", "generic_name": "Omeprazole", "brand": "Ferozsons", "category": "Gastrology", "form": "Capsule", "strength": "20mg", "unit": "Strip", "pack_size": 14, "manufacturer": "Ferozsons", "requires_prescription": True, "mrp": 180.0, "reorder_point": 25, "min_stock_level": 10},
        ]

        med_objects = []
        for m in medicines_data:
            med = Medicine(**m)
            db.add(med)
            med_objects.append(med)
        await db.flush()

        # Create batches with stock
        from app.models.inventory import MedicineBatch
        from datetime import date, timedelta
        import random

        for med in med_objects:
            batch = MedicineBatch(
                medicine_id=med.id,
                branch_id=main_branch.id,
                batch_number=f"BTH-{med.name[:3].upper()}-2024",
                expiry_date=date.today() + timedelta(days=random.randint(180, 730)),
                manufacturing_date=date.today() - timedelta(days=90),
                quantity=random.randint(50, 200),
                purchase_price=float(med.mrp) * 0.7 if med.mrp else 100,
                selling_price=float(med.mrp) if med.mrp else 150,
            )
            db.add(batch)
        await db.flush()

        # Create patients
        patients_data = [
            {"name": "Muhammad Ali Khan", "phone": "0300-1234567", "cnic": "42201-1234567-1", "gender": "Male", "blood_group": "A+", "customer_type": "retail"},
            {"name": "Fatima Zahra Siddiqui", "phone": "0321-7654321", "cnic": "42201-7654321-2", "gender": "Female", "blood_group": "O+", "customer_type": "retail", "chronic_conditions": ["Diabetes", "Hypertension"]},
            {"name": "Ahmed Hassan", "phone": "0333-1111222", "gender": "Male", "customer_type": "hospital", "ward": "Cardiology"},
        ]

        for p_data in patients_data:
            patient = Patient(branch_id=main_branch.id, **p_data)
            db.add(patient)
        await db.flush()

        # Create a supplier
        from app.models.supplier import Supplier
        supplier = Supplier(
            name="Getz Pharma Pvt Ltd",
            contact_person="Sales Manager",
            phone="+92-21-111-438-9723",
            email="orders@getzpharma.com",
            address="Korangi Industrial Area, Karachi",
            credit_days=30,
            payment_terms="Net 30",
            is_active=True,
        )
        db.add(supplier)

        # System settings
        from app.models.settings import SystemSettings
        default_settings = [
            {"key": "pharmacy_name", "value": {"v": "RxPOS Pharmacy"}, "description": "Pharmacy name for receipts"},
            {"key": "tax_rate", "value": {"v": 0.17}, "description": "Default GST rate"},
            {"key": "currency", "value": {"v": "PKR"}, "description": "Currency"},
            {"key": "low_stock_alert_days", "value": {"v": 30}, "description": "Alert when stock covers < X days"},
            {"key": "expiry_alert_days", "value": {"v": [30, 60, 90]}, "description": "Alert at X days before expiry"},
        ]
        for s in default_settings:
            db.add(SystemSettings(**s))

        await db.commit()
        print("✅ Seed complete!")
        print("  Users: admin@rxpos.pk (admin123), cashier@rxpos.pk (cashier123), pharmacist@rxpos.pk (cashier123), manager@rxpos.pk (manager123)")
        print("  Branches: HQ (Main), NB (North)")
        print(f"  Medicines: {len(medicines_data)}")
        print(f"  Patients: {len(patients_data)}")

    await engine.dispose()


if __name__ == "__main__":
    asyncio.run(seed())

File: scripts/src/seed.ts
import { db } from "@workspace/db";
import {
  branchesTable,
  usersTable,
  medicinesTable,
  batchesTable,
  patientsTable,
  suppliersTable,
  prescriptionsTable,
} from "@workspace/db";
import bcrypt from "bcryptjs";
import { eq } from "drizzle-orm";

async function seed() {
  console.log("Seeding database...");

  // Upsert branches
  await db.insert(branchesTable).values({
    name: "Main Dispensary",
    code: "MAIN",
    address: "123 Hospital Road, Lahore",
    phone: "042-35761234",
    licenseNumber: "DL-2024-001",
    taxRate: "0.17",
    mrpEnforcement: true,
  }).onConflictDoNothing();

  await db.insert(branchesTable).values({
    name: "Gulberg Branch",
    code: "GLB",
    address: "45 Gulberg Main Blvd, Lahore",
    phone: "042-35880099",
    licenseNumber: "DL-2024-002",
    taxRate: "0.17",
    mrpEnforcement: true,
  }).onConflictDoNothing();

  const [branch] = await db.select().from(branchesTable).where(eq(branchesTable.code, "MAIN"));
  const [branch2] = await db.select().from(branchesTable).where(eq(branchesTable.code, "GLB"));

  const hash = await bcrypt.hash("admin123", 12);
  const cashierHash = await bcrypt.hash("cashier123", 12);

  const [admin] = await db.insert(usersTable).values({
    name: "Dr. Admin User",
    email: "admin@rxpos.pk",
    passwordHash: hash,
    role: "admin",
    branchId: branch.id,
    phone: "0321-4567890",
    permissions: ["all"],
  }).returning();

  await db.insert(usersTable).values({
    name: "Sarah Khan",
    email: "cashier@rxpos.pk",
    passwordHash: cashierHash,
    role: "cashier",
    branchId: branch.id,
    phone: "0322-1234567",
    permissions: ["sales.create", "patients.view"],
  });

  await db.insert(usersTable).values({
    name: "Ali Hassan",
    email: "pharmacist@rxpos.pk",
    passwordHash: cashierHash,
    role: "pharmacist",
    branchId: branch.id,
    phone: "0333-9876543",
    permissions: ["prescriptions.verify", "prescriptions.dispense", "inventory.adjust"],
  });

  const medicines = await db.insert(medicinesTable).values([
    {
      name: "Paracetamol 500mg",
      genericName: "Paracetamol",
      brand: "Panadol",
      category: "Analgesics",
      manufacturer: "GSK Pakistan",
      strength: "500mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 20,
      mrp: "45.00",
      requiresPrescription: false,
      isEssentialMedicine: true,
      minStockLevel: 50,
      reorderPoint: 100,
      barcode: "6009875021001",
    },
    {
      name: "Amoxicillin 500mg",
      genericName: "Amoxicillin",
      brand: "Amoxil",
      category: "Antibiotics",
      manufacturer: "GSK Pakistan",
      strength: "500mg",
      form: "Capsule",
      unit: "Cap",
      packSize: 14,
      mrp: "185.00",
      requiresPrescription: true,
      isEssentialMedicine: true,
      minStockLevel: 20,
      reorderPoint: 40,
      barcode: "6009875021002",
    },
    {
      name: "Metformin 500mg",
      genericName: "Metformin HCl",
      brand: "Glucophage",
      category: "Antidiabetics",
      manufacturer: "Merck Pakistan",
      strength: "500mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 30,
      mrp: "320.00",
      requiresPrescription: true,
      isEssentialMedicine: true,
      minStockLevel: 30,
      reorderPoint: 60,
    },
    {
      name: "Omeprazole 20mg",
      genericName: "Omeprazole",
      brand: "Losec",
      category: "GI Drugs",
      manufacturer: "AstraZeneca",
      strength: "20mg",
      form: "Capsule",
      unit: "Cap",
      packSize: 14,
      mrp: "155.00",
      requiresPrescription: false,
      minStockLevel: 25,
      reorderPoint: 50,
    },
    {
      name: "Amlodipine 5mg",
      genericName: "Amlodipine",
      brand: "Norvasc",
      category: "Cardiovascular",
      manufacturer: "Pfizer Pakistan",
      strength: "5mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 30,
      mrp: "285.00",
      requiresPrescription: true,
      minStockLevel: 20,
      reorderPoint: 40,
    },
    {
      name: "Cetirizine 10mg",
      genericName: "Cetirizine HCl",
      brand: "Zyrtec",
      category: "Antihistamines",
      manufacturer: "UCB",
      strength: "10mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 10,
      mrp: "120.00",
      requiresPrescription: false,
      minStockLevel: 40,
      reorderPoint: 80,
    },
    {
      name: "Ibuprofen 400mg",
      genericName: "Ibuprofen",
      brand: "Brufen",
      category: "NSAIDs",
      manufacturer: "Abbott Pakistan",
      strength: "400mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 20,
      mrp: "95.00",
      requiresPrescription: false,
      isEssentialMedicine: true,
      minStockLevel: 40,
      reorderPoint: 80,
    },
    {
      name: "Atorvastatin 20mg",
      genericName: "Atorvastatin",
      brand: "Lipitor",
      category: "Cardiovascular",
      manufacturer: "Pfizer Pakistan",
      strength: "20mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 30,
      mrp: "450.00",
      requiresPrescription: true,
      minStockLevel: 15,
      reorderPoint: 30,
    },
    {
      name: "Azithromycin 500mg",
      genericName: "Azithromycin",
      brand: "Zithromax",
      category: "Antibiotics",
      manufacturer: "Pfizer Pakistan",
      strength: "500mg",
      form: "Tablet",
      unit: "Tab",
      packSize: 3,
      mrp: "380.00",
      requiresPrescription: true,
      minStockLevel: 10,
      reorderPoint: 20,
    },
    {
      name: "Salbutamol Inhaler",
      genericName: "Salbutamol",
      brand: "Ventolin",
      category: "Respiratory",
      manufacturer: "GSK Pakistan",
      strength: "100mcg",
      form: "Inhaler",
      unit: "Pcs",
      packSize: 1,
      mrp: "320.00",
      requiresPrescription: true,
      minStockLevel: 10,
      reorderPoint: 20,
    },
  ]).returning();

  const today = new Date().toISOString().split("T")[0];
  const future1 = new Date();
  future1.setFullYear(future1.getFullYear() + 1);
  const future1Str = future1.toISOString().split("T")[0];
  const future2 = new Date();
  future2.setFullYear(future2.getFullYear() + 2);
  const future2Str = future2.toISOString().split("T")[0];
  const expiring = new Date();
  expiring.setDate(expiring.getDate() + 45);
  const expiringStr = expiring.toISOString().split("T")[0];

  for (let i = 0; i < medicines.length; i++) {
    const med = medicines[i];
    await db.insert(batchesTable).values({
      medicineId: med.id,
      branchId: branch.id,
      batchNumber: `BT-${String(i + 1).padStart(4, "0")}-A`,
      expiryDate: i % 3 === 0 ? expiringStr : i % 2 === 0 ? future1Str : future2Str,
      quantity: Math.floor(Math.random() * 150) + 20,
      purchasePrice: String((Number(med.mrp) * 0.65).toFixed(2)),
      sellingPrice: med.mrp,
    });
  }

  const [supplier] = await db.insert(suppliersTable).values({
    name: "MediPharma Distributors",
    contactPerson: "Tariq Mehmood",
    phone: "0300-4581234",
    email: "tariq@medipharma.pk",
    address: "Hafeez Centre, Lahore",
    licenseNumber: "WDL-2024-045",
    creditDays: 30,
    paymentTerms: "Net 30",
  }).returning();

  await db.insert(suppliersTable).values({
    name: "National Health Supplies",
    contactPerson: "Rashida Bibi",
    phone: "0321-9078654",
    email: "rashida@nhs.pk",
    address: "Shadman, Lahore",
    creditDays: 45,
    paymentTerms: "Net 45",
  });

  const [patient1] = await db.insert(patientsTable).values({
    name: "Muhammad Imran",
    phone: "0345-6789012",
    cnic: "35202-1234567-1",
    gender: "Male",
    dob: "1975-03-15",
    address: "DHA Phase 5, Lahore",
    allergies: ["Penicillin", "Sulfa"],
    chronicConditions: ["Hypertension", "Type 2 Diabetes"],
    chronicMedications: ["Metformin 500mg", "Amlodipine 5mg"],
    customerType: "retail",
    bloodGroup: "B+",
  }).returning();

  const [patient2] = await db.insert(patientsTable).values({
    name: "Fatima Malik",
    phone: "0333-2345678",
    cnic: "35202-7654321-0",
    gender: "Female",
    dob: "1990-07-22",
    address: "Gulberg III, Lahore",
    allergies: [],
    chronicConditions: ["Asthma"],
    chronicMedications: ["Salbutamol Inhaler"],
    customerType: "retail",
    bloodGroup: "A+",
  }).returning();

  await db.insert(patientsTable).values({
    name: "Ahmed Raza",
    phone: "0312-8765432",
    mrn: "MRN-2024-001",
    ward: "Cardiology",
    gender: "Male",
    dob: "1960-11-08",
    allergies: ["Aspirin"],
    chronicConditions: ["Coronary Artery Disease"],
    chronicMedications: ["Atorvastatin 20mg"],
    customerType: "hospital",
    bloodGroup: "O+",
  });

  await db.insert(prescriptionsTable).values([
    {
      patientId: patient1.id,
      prescriberName: "Dr. Khalid Mahmood",
      prescriberLicense: "PMDC-12345",
      prescriptionDate: today,
      status: "received",
      source: "manual",
      items: [
        { medicine_name: "Metformin 500mg", dose: "500mg", frequency: "BD", duration: "30 days", quantity: 60 },
        { medicine_name: "Amlodipine 5mg", dose: "5mg", frequency: "OD", duration: "30 days", quantity: 30 },
      ],
      branchId: branch.id,
      notes: "Patient on long-term diabetes management",
    },
    {
      patientId: patient2.id,
      prescriberName: "Dr. Sana Mirza",
      prescriberLicense: "PMDC-67890",
      prescriptionDate: today,
      status: "verified",
      source: "manual",
      items: [
        { medicine_name: "Salbutamol Inhaler", dose: "100mcg", frequency: "PRN", duration: "60 days", quantity: 2 },
        { medicine_name: "Cetirizine 10mg", dose: "10mg", frequency: "OD", duration: "7 days", quantity: 7 },
      ],
      branchId: branch.id,
      notes: "Seasonal allergic asthma exacerbation",
    },
  ]);

  console.log("Seeding complete!");
  console.log("Login credentials:");
  console.log("  Admin:      admin@rxpos.pk / admin123");
  console.log("  Cashier:    cashier@rxpos.pk / cashier123");
  console.log("  Pharmacist: pharmacist@rxpos.pk / cashier123");
  process.exit(0);
}

seed().catch((err) => {
  console.error("Seed failed:", err);
  process.exit(1);
});
